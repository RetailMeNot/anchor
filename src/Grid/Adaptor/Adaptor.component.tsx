// REACT
import * as React from 'react';
// VENDOR
import Responsive from 'react-responsive';
// COMPONENTS
import { ResponsiveContext } from '../ResponsiveProvider';
import { getMinMax } from '../utils';

interface CustomAdaptorProps {
    children?: any;
    query?: string;
    minHeight?: number | string;
    maxHeight?: number | string;
    minDeviceHeight?: number | string;
    maxDeviceHeight?: number | string;
    minWidth?: number | string;
    maxWidth?: number | string;
    minDeviceWidth?: number | string;
    maxDeviceWidth?: number | string;
    minResolution?: number | string;
    maxResolution?: number | string;
    onBeforeChange?: (matches: boolean) => void;
    onChange?: (matches: boolean) => void;
}
interface FixedAdaptor extends CustomAdaptorProps {
    maxWidth?: never;
    minWidth?: never;
    minDeviceWidth?: never;
    maxDeviceWidth?: never;
}

export const CustomAdaptor = ({ ...props }: CustomAdaptorProps) => (
    <Responsive className="anchor-adaptor" {...props} />
);

export const XS = ({ ...props }: FixedAdaptor) => {
    const { breakpoints } = React.useContext(ResponsiveContext);
    const { max } = getMinMax(0, breakpoints);

    return max !== 0 ? (
        <Responsive className="anchor-adaptor" {...props} maxWidth={max} />
    ) : null;
};

export const SM = ({ ...props }: FixedAdaptor) => {
    const { breakpoints } = React.useContext(ResponsiveContext);
    const { min, max } = getMinMax(1, breakpoints);

    return min !== 0 && max !== 0 ? (
        <Responsive
            className="anchor-adaptor"
            {...props}
            minWidth={min}
            maxWidth={max}
        />
    ) : null;
};

export const MD = ({ ...props }: FixedAdaptor) => {
    const { breakpoints } = React.useContext(ResponsiveContext);
    const { min, max } = getMinMax(2, breakpoints);

    return min !== 0 && max !== 0 ? (
        <Responsive
            className="anchor-adaptor"
            {...props}
            minWidth={min}
            maxWidth={max}
        />
    ) : null;
};

export const LG = ({ ...props }: FixedAdaptor) => {
    const { breakpoints } = React.useContext(ResponsiveContext);
    const { min, max } = getMinMax(3, breakpoints);

    return min !== 0 && max !== 0 ? (
        <Responsive
            className="anchor-adaptor"
            {...props}
            minWidth={min}
            maxWidth={max}
        />
    ) : null;
};

export const XL = ({ ...props }: FixedAdaptor) => {
    const { breakpoints } = React.useContext(ResponsiveContext);
    const { min, max } = getMinMax(4, breakpoints);

    return min !== 0 && max !== 0 ? (
        <Responsive
            className="anchor-adaptor"
            {...props}
            minWidth={min}
            maxWidth={max}
        />
    ) : null;
};

export const XXL = ({ ...props }: FixedAdaptor) => {
    const { breakpoints } = React.useContext(ResponsiveContext);
    const { min } = getMinMax(5, breakpoints);

    return min !== 0 ? (
        <Responsive className="anchor-adaptor" {...props} minWidth={min} />
    ) : null;
};
