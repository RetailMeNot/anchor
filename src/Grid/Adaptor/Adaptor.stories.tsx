// REACT
import * as React from 'react';
// STORYBOOK
import { storiesOf } from '@storybook/react';
import { boolean, number } from '@storybook/addon-knobs';
// VENDOR
import styled from '@xstyled/styled-components';
import { ThemeProvider } from '@xstyled/styled-components';
// COMPONENT
import { XS, SM, MD, LG, XL, XXL, CustomAdaptor } from './Adaptor.component';
import { ResponsiveProvider } from '../ResponsiveProvider';
import { RootTheme } from '../../theme';
import { Typography } from '../../index';
// README
import * as README from './README.md';

const StyledStory = styled('div')`
    padding: 2rem 5rem;
`;

storiesOf('Components/Grid/Adaptor', module)
    .addParameters({
        readme: {
            sidebar: README,
        },
    })
    .add('Custom Breakpoints', () => (
        <ThemeProvider theme={RootTheme}>
            <StyledStory>
                <h4>Use knobs to set custom breakpoints</h4>
                <CustomAdaptor
                    minWidth={number('Minimum Width', 1)}
                    maxWidth={number('Maximum Width', 99999)}
                >
                    <p>My dimensions are defined by knobs.</p>
                </CustomAdaptor>
            </StyledStory>
        </ThemeProvider>
    ))
    .add('Fixed Adaptors', () => {
        return (
            <ThemeProvider theme={RootTheme}>
                <ResponsiveProvider
                    debug={boolean('ResponsiveProvider Debug', true)}
                >
                    <StyledStory>
                        <XS>
                            <Typography tag="h4">XS</Typography>
                        </XS>
                        <SM>
                            <Typography tag="h4">SM</Typography>
                        </SM>
                        <MD>
                            <Typography tag="h4">MD</Typography>
                        </MD>
                        <LG>
                            <Typography tag="h4">LG</Typography>
                        </LG>
                        <XL>
                            <Typography tag="h4">XL</Typography>
                        </XL>
                        <XXL>
                            <Typography tag="h4">XXL</Typography>
                        </XXL>
                    </StyledStory>
                </ResponsiveProvider>
            </ThemeProvider>
        );
    });
