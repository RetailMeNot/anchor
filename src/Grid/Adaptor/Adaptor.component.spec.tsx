// REACT
import * as React from 'react';
import * as renderer from 'react-test-renderer';
// COMPONENT
import { CustomAdaptor } from './Adaptor.component';
// ENZYME
import { shallow, mount } from 'enzyme';

// TEST SETUP
const subject = <CustomAdaptor minWidth={200} />;
const wrapper = mount(subject);
const component = shallow(subject);

describe('Component: Adaptor', () => {
    it('should be defined', () => {
        expect(subject).toBeDefined();
        expect(wrapper).toBeDefined();
        expect(component).toBeDefined();
        const tree = renderer.create(subject).toJSON();

        expect(tree).toMatchSnapshot();
    });
});
