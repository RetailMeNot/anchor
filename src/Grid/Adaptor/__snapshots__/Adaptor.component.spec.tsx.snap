// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component: Adaptor should be defined 1`] = `null`;

exports[`Component: LG should be defined 1`] = `null`;

exports[`Component: MD should be defined 1`] = `null`;

exports[`Component: SM should be defined 1`] = `null`;

exports[`Component: XL should be defined 1`] = `null`;

exports[`Component: XS should be defined 1`] = `null`;

exports[`Component: XXL should be defined 1`] = `null`;
